File: 2020/go/day05/day05.go
package main

import(
	"fmt"
	"os"
	"bufio"
	"strconv"
	"sort"
	"strings"
)

func main() {
	file, err := os.Open("/Users/hhh/JungleGym/advent_of_code/2020/inputs/day05.in")
	if err != nil {
		panic(err)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)
	seats := []string{} // binary rep of seats
	r := strings.NewReplacer("F", "0", "B", "1", "L", "0", "R", "1")

	for scanner.Scan() {
		line := scanner.Text()
		b := r.Replace(line)
		seats = append(seats, b)
	}
	sort.Strings(seats)
	max, _ := strconv.ParseInt(seats[len(seats)-1], 2, 11)
	fmt.Printf("part 1: %d\n", max)

	// Part 2
	offset, _ := strconv.ParseInt(seats[0], 2, 11)
	for i, s := range seats {
		seat, _ := strconv.ParseInt(s, 2, 11)
		if seat != int64(i) + offset {
			fmt.Printf("part 1: %d\n", seat - 1)
			break
		}
	}
}

File: 2020/day05.rb
input = File.readlines("/Users/hhh/JungleGym/advent_of_code/2020/inputs/day05.in").map { |x| x.strip }

# map each line to binary representation
seats = input.map do |x|
  x.tr("FBLR", "0101")
end

# binary rep already is the same thing as multiplying row by 8 and adding column,
# since we're shifting the first 7 bits left by 3 bits
id = seats.max
p "part 1: #{id.to_i(2)}"

# to find missing seat, find which of the sorted list does not correspond to
# index plus offset of first seat
sorted = seats.map do |b|
  b.to_i(2)
end.sort

offset = sorted.first

sorted.each_with_index do |s, i|
  if s != i + offset
    p "part 2: #{s - 1}"
    break
  end
  i += 1
end
